### PR TITLE
ENH: DICOM To NRRD Module Implementation

### PR DESCRIPTION
**Pull request by alex-myers**
The DICOM To NRRD module provides batch processing for converting DICOM files into NRRD files and outputs the files into a DICOM standard conforming folder hierarchy of Patient/Study/Series.

Reference discussion forum:
https://discourse.slicer.org/t/
possible-new-slicer-extension-dicomtonrrdbatchprocessing/33284

### CODE REVIEW

**Review by lassoan — COMMENTED** [review]
Thank you for your contribution. It would be useful to make batch conversion from DICOM to research file formats easier and adding a module like this could be a sensible approach.

Since the module will need some time to mature (go through more testing, get more features, etc.) we should add it to an extension first. We could use the [Sandbox](https://github.com/PerkLab/SlicerSandbox).

File: Modules/Scripted/DICOMToNRRD/DICOMToNRRD.py [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+            secondPixelSpacing = pixelSpacing[slashIndexPlusOne:]
+            pixSpaceStrList = firstPixelSpacing + "," + secondPixelSpacing
+
+        return pixSpaceStrList
+
+    # Save DICOM information and write a PatientDICOMInfo.json for a patient series
+    def SaveJsonFile(self, patientSeriesDir, pixSpaceStrList, instUid):
+        # Generate a dictionary of patient DICOM information for the series
+        dictionary = {
+            # DICOM header meta information
```
[the remark this answers is not in the document]
Comment (alex-myers) [reply Modules/Scripted/DICOMToNRRD/DICOMToNRRD.py:157]: Valid point. I'll remove the FetchPixelSpacingString method and keep the default.

File: Modules/Scripted/DICOMToNRRD/DICOMToNRRD.py [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+        # Write JSON file
+        with open(jsonFile, 'w') as f:
+            f.write(json_object)
+
+    # Create the patient series directory only if it exists
+    def CreatePatientSeriesDirectory(self, patientSeriesDir):
+        if not os.path.exists(patientSeriesDir):
+            os.makedirs(patientSeriesDir)
+
+    # Create a patient name directory based on patient name and UID in case you have patients with the same first and last names
```
[the remark this answers is not in the document]
Comment (alex-myers) [reply Modules/Scripted/DICOMToNRRD/DICOMToNRRD.py:232]: Ahh, thanks for bringing this to my attention. Docutil won't pick up those comments: https://peps.python.org/pep-0257/#one-line-docstrings 

I'll update those changes.

File: Modules/Scripted/DICOMToNRRD/DICOMToNRRD.py [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+        if studyText == "":
+            studyText = shNode.GetItemAttribute(studyItem, 'DICOM.StudyDescription')
+        # Study Description comes back with "No study description" when there is no study
+        return studyText
+
+    # Formulate a patient series name
+    def PatientSeriesName(self, shNode, seriesItem):
+        seriesName = shNode.GetItemAttribute(seriesItem, 'DICOM.SeriesNumber')
+        if seriesName == "":
+            seriesName = shNode.GetItemAttribute(seriesItem, 'DICOM.SeriesDescription')
```
[the remark this answers is not in the document]
Comment (alex-myers) [reply Modules/Scripted/DICOMToNRRD/DICOMToNRRD.py:255]: I'll put in logic so only alpha numeric characters are generated

File: Modules/Scripted/DICOMToNRRD/DICOMToNRRD.py [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+            patientName = "Patient " + DateNowPrettyFormat()
+        return patientName
+
+    # Compose a study name for the patient
+    def PatientStudyName(self, shNode, studyItem):
+        studyText = shNode.GetItemAttribute(studyItem, 'DICOM.StudyDate')
+        if studyText == "":
+            studyText = shNode.GetItemAttribute(studyItem, 'DICOM.StudyID')
+        if studyText == "":
+            studyText = shNode.GetItemAttribute(studyItem, 'DICOM.StudyDescription')
```
[the remark this answers is not in the document]
Comment (alex-myers) [reply Modules/Scripted/DICOMToNRRD/DICOMToNRRD.py:247]: I'll put in logic so only alpha numeric characters are generated

File: Modules/Scripted/DICOMToNRRD/DICOMToNRRD.py [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+            DICOMUtils.importDicom(dicomDataDir, db)  # load the DICOM files into a Slicer DB
+            patientUIDs = db.patients()
+            for patientUID in patientUIDs:
+                try:
+                    loadedNodeIDs = DICOMUtils.loadPatientByUID(patientUID)
+                    for loadedNodeID in loadedNodeIDs:
+                        # Check if we want to save this node
+                        node = slicer.mrmlScene.GetNodeByID(loadedNodeID)
+                        # Only export images that are volumes
+                        if not node or not node.IsA('vtkMRMLScalarVolumeNode'):
```
[the remark this answers is not in the document]
Comment (alex-myers) [reply Modules/Scripted/DICOMToNRRD/DICOMToNRRD.py:282]: I'll add documentation stating this is only for scalar volumes and not segmentations, structure sets etc. Should I rename the module to DICOM To NRRD SV?

**Comment by alex-myers** [comment]
Before the next pull request, I'll test out with the 40,000+ patients.

**Pull request closed without merging**